### PR TITLE
Python: Support more dictionary read/store steps

### PR DESCRIPTION
**Pull request by RasmusWL**

### CODE REVIEW

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -615,17 +615,38 @@
...
+  or
+  exists(SubscriptNode subscript |
+    nodeTo.(PostUpdateNode).getPreUpdateNode().asCfgNode() = subscript.getObject() and
+    nodeFrom.asCfgNode() = subscript.(DefinitionNode).getValue() and
+    c.getKey() = subscript.getIndex().getNode().(StrConst).getText()
+  )
+  or
+  // see https://docs.python.org/3.10/library/stdtypes.html#dict.setdefault
+  exists(MethodCallNode call |
+    call.calls(nodeTo.(PostUpdateNode).getPreUpdateNode(), ["setdefault"]) and
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:638]:
```suggestion
    call.calls(nodeTo.(PostUpdateNode).getPreUpdateNode(), "setdefault") and
```

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -615,17 +615,38 @@
...
+    call.getArg(0).asExpr().(StrConst).getText() = c.(DictionaryElementContent).getKey() and
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:639]:
```suggestion
    call.getArg(0).asExpr().(StrConst).getText() = c.getKey() and
```

**Review by RasmusWL — COMMENTED** [review]
Fixes for ql4ql alerts

**Review by yoff — DISMISSED** [review]
Nice PR, good tests, everything looks as it should so I will approve. The evaluation is a bit of a mixed bag, though, so we could entertain the option of considering this blocked on fixing some of the unearthed issues...apart from the `saltstack` results, it does not look like a big explosion, though..

File: python/ql/test/experimental/dataflow/fieldflow/UnresolvedCalls.ql [hunk 3]
```diff
@@ -1,2 +1,12 @@
...
 import experimental.dataflow.TestUtil.UnresolvedCalls
+private import semmle.python.dataflow.new.DataFlow
+
+class IgnoreDictMethod extends UnresolvedCallExpectations {
+  override predicate unresolvedCall(CallNode call) {
+    super.unresolvedCall(call) and
+    not any(DataFlow::MethodCallNode methodCall |
+      methodCall.getMethodName() in ["get", "setdefault"]
+    ).asCfgNode() = call
+  }
```
Comment (yoff) [inline python/ql/test/experimental/dataflow/fieldflow/UnresolvedCalls.ql:11]: Nice!

**Review by yoff — APPROVED** [review]
Great to have these flows now!

**Comment by RasmusWL** [comment]
Performance and result changes looks good :+1:

**Pull request merged**